#VRML_SIM R2025a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/floors/protos/Floor.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/apartment_structure/protos/Wall.proto"

WorldInfo {
  basicTimeStep 5
  FPS 120
  contactProperties [
    ContactProperties {
    }
  ]
}
Viewpoint {
  orientation -0.3501733466570501 0.5119753797996859 0.7843850060843647 1.4342298902532036
  position 4.0296836151773014 -25.246161571066207 40.09743864578677
  follow "robot"
  followType "None"
}
TexturedBackground {
}
TexturedBackgroundLight {
}
DEF DRONE Robot {
  translation 1.25 1.25 0.219264
  children [
    Camera {
      translation 0 0.14 0.03
      rotation 0 0 1 1.5708
      fieldOfView 2
      width 128
      antiAliasing TRUE
    }
    Altimeter {
    }
    Accelerometer {
      name "acc"
    }
    GPS {
    }
    LED {
      translation 0 -0.076933 -0.022
      children [
        Shape {
          appearance PBRAppearance {
            roughness 0.5
            metalness 0
            emissiveIntensity 200
          }
          geometry Box {
            size 0.01 0.015 0.008
          }
        }
      ]
      color [
        0 1 0
      ]
    }
    Solid {
      rotation 1 0 0 1.57079632675
      children [
        Shape {
          appearance Appearance {
            material Material {
              diffuseColor 0.2 0.2 0.2
            }
          }
          geometry Mesh {
            url [
              "../protos/meshes/M100.stl"
            ]
          }
        }
      ]
      boundingObject Box {
        size 0.1 0.1 0.1
      }
    }
    InertialUnit {
      name "imu"
    }
    Gyro {
    }
    Propeller {
      shaftAxis 0 0 1
      centerOfThrust 0.23 0.23 0.03
      thrustConstants 2.55e-05 0
      torqueConstants 5.1e-06 0
      fastHelixThreshold 130
      device RotationalMotor {
        name "motor4"
        maxVelocity 5000
        maxTorque 5000
      }
      fastHelix Solid {
        translation 0.23 0.23 0.03
        children [
          Shape {
            appearance PBRAppearance {
              baseColor 0.5 0.5 0.5
              baseColorMap ImageTexture {
                url [
                  "../protos/textures/fast_helix.png"
                ]
              }
              metalness 0
            }
            geometry Cylinder {
              height 0.01
              radius 0.14
            }
          }
        ]
        name "solid(1)"
      }
      slowHelix Solid {
        translation 0.23 0.23 0.03
        children [
          Solid {
            children [
              Shape {
                appearance PBRAppearance {
                  baseColor 0.5 0.5 0.5
                  roughness 0.9
                  metalness 0
                }
                geometry Mesh {
                  url [
                    "../protos/meshes/propeller_ccw.dae"
                  ]
                }
              }
            ]
          }
        ]
      }
    }
    Propeller {
      shaftAxis 0 0 1
      centerOfThrust 0.23 -0.23 0.03
      thrustConstants -2.55e-05 0
      torqueConstants 5.1e-06 0
      fastHelixThreshold 130
      device RotationalMotor {
        name "motor3"
        maxVelocity 5000
        maxTorque 5000
      }
      fastHelix Solid {
        translation 0.23 -0.23 0.03
        children [
          Shape {
            appearance PBRAppearance {
              baseColor 0.5 0.5 0.5
              baseColorMap ImageTexture {
                url [
                  "../protos/textures/fast_helix.png"
                ]
              }
              metalness 0
            }
            geometry Cylinder {
              height 0.01
              radius 0.14
            }
          }
        ]
        name "solid(1)"
      }
      slowHelix Solid {
        translation 0.23 -0.23 0.03
        rotation 0 0 -1 0
        children [
          Solid {
            children [
              Shape {
                appearance PBRAppearance {
                  baseColor 0.5 0.5 0.5
                  roughness 0.9
                  metalness 0
                }
                geometry Mesh {
                  url [
                    "../protos/meshes/propeller_cw.dae"
                  ]
                }
              }
            ]
          }
        ]
      }
    }
    Propeller {
      shaftAxis 0 0 1
      centerOfThrust -0.23 -0.23 0.03
      thrustConstants 2.55e-05 0
      torqueConstants 5.1e-06 0
      fastHelixThreshold 130
      device RotationalMotor {
        name "motor2"
        maxVelocity 5000
        maxTorque 5000
      }
      fastHelix Solid {
        translation -0.23 -0.23 0.03
        children [
          Shape {
            appearance PBRAppearance {
              baseColor 0.5 0.5 0.5
              baseColorMap ImageTexture {
                url [
                  "../protos/textures/fast_helix.png"
                ]
              }
              metalness 0
            }
            geometry Cylinder {
              height 0.01
              radius 0.14
            }
          }
        ]
        name "solid(1)"
      }
      slowHelix Solid {
        translation -0.23 -0.23 0.03
        children [
          Solid {
            children [
              Shape {
                appearance PBRAppearance {
                  baseColor 0.5 0.5 0.5
                  roughness 0.9
                  metalness 0
                }
                geometry Mesh {
                  url [
                    "../protos/meshes/propeller_ccw.dae"
                  ]
                }
              }
            ]
          }
        ]
      }
    }
    Propeller {
      shaftAxis 0 0 1
      centerOfThrust -0.23 0.23 0.03
      thrustConstants -2.55e-05 0
      torqueConstants 5.1e-06 0
      fastHelixThreshold 130
      device RotationalMotor {
        name "motor1"
        maxVelocity 5000
        maxTorque 5000
      }
      fastHelix Solid {
        translation -0.23 0.23 0.03
        children [
          Shape {
            appearance PBRAppearance {
              baseColor 0.5 0.5 0.5
              baseColorMap ImageTexture {
                url [
                  "../protos/textures/fast_helix.png"
                ]
              }
              metalness 0
            }
            geometry Cylinder {
              height 0.01
              radius 0.14
            }
          }
        ]
        name "solid(1)"
      }
      slowHelix Solid {
        translation -0.23 0.23 0.03
        rotation 0 0 -1 0
        children [
          Solid {
            children [
              Shape {
                appearance PBRAppearance {
                  baseColor 0.5 0.5 0.5
                  roughness 0.9
                  metalness 0
                }
                geometry Mesh {
                  url [
                    "../protos/meshes/propeller_cw.dae"
                  ]
                }
              }
            ]
          }
        ]
      }
    }
  ]
  boundingObject Box {
    size 0.5 0.5 0.4
  }
  physics Physics {
    mass 3
  }
  controller "drone_controller"
  supervisor TRUE
}
Wall {
  translation 5 5 0
  size 10 1 3
}
Wall {
  translation 5 18 0
  name "wall(10)"
  size 10 1 3
}
Wall {
  translation 27.5 7 0
  name "wall(11)"
  size 5 1 3
}
Wall {
  translation 17 10 0
  name "wall(7)"
  size 10 1 3
}
Wall {
  translation 20 5 0
  name "wall(5)"
  size 1 10 3
}
Wall {
  translation 10 17 0
  name "wall(2)"
  size 1 6 3
}
Wall {
  translation 14 16 1
  name "wall(3)"
  size 1 1 3
}
Wall {
  translation 14 16 1
  name "wall(8)"
  size 1 1 3
}
Wall {
  translation 18 16 1
  name "wall(9)"
  size 1 1 3
}
Wall {
  translation 22 25 1
  name "wall(16)"
  size 1 1 3
}
Wall {
  translation 20 20 0
  name "wall(4)"
  size 2 2 3
}
Wall {
  translation 8 25 0
  name "wall(12)"
  size 2 2 3
}
Wall {
  translation 15 22 0
  name "wall(6)"
  size 2 2 3
}
Wall {
  translation 26 22 0
  name "wall(13)"
  size 2 2 3
}
Wall {
  translation 24 15 0
  name "wall(14)"
  size 2 2 3
}
Wall {
  translation 24 15 0
  name "wall(15)"
  size 2 2 3
}
Floor {
  translation 15 15 0
  size 30 30
  tileSize 100 100
}
Robot {
  name "superviewer"
  controller "trajectory_supervisor"
  supervisor TRUE
}
Robot {
  translation 0.1 0 0.1
  name "superviewer_point"
  controller "path_supervisor"
  supervisor TRUE
}
